### ISSUE TITLE
Invoke local with custom runtime fails because unable to unzip whole package

### ISSUE THREAD

**Issue by marcinczeczko** [serverless/serverless#6028]
# This is a Bug Report
Invoke local with docker fails because of ZIP with folders

## Description
* What went wrong?
When running `sls invoke local --docker -f hello` I'm getting the error
```
Error: ENOENT: no such file or directory, open '.serverless/invokeLocal/artifact/lib/org.jboss.slf4j.slf4j-jboss-logging-1.1.0.Final.jar'
    at zip.files.(anonymous function).async.then.fileData (/usr/local/lib/node_modules/serverless/lib/plugins/aws/invokeLocal/index.js:309:23)
    at runNextTicks (internal/process/next_tick.js:47:5)
    at processImmediate (timers.js:610:7)
    at process.topLevelDomainCallback (domain.js:126:23)
```
It's due to the fact the function zip consists a subfolders, the zip contents is as follows:
```
├── amazon-lambda-dynamodb-1.0-SNAPSHOT-runner.jar
├── bootstrap
└── lib
    ├── some.jar
    ├── other.jar
```
If checked `.serverless` folder you can find that only top level jar & bootstrap file is extracted
```
.serverless
├── invokeLocal
│   ├── Dockerfile
│   └── artifact
│       ├── amazon-lambda-dynamodb-1.0-SNAPSHOT-runner.jar
│       └── bootstrap
```
* What did you expect should have happened?
Zip is extracted and docker started

* What was the config you used?
```
service: hello-lambda

provider:
  name: aws
  runtime: provided
  stage: dev
  region: eu-central-1

package:
  artifact: target/amazon-lambda-dynamodb-1.0-SNAPSHOT.zip

functions:
  hello:
    handler: handler
    events:
      - http:
          method: get
          path: user
```
* What stacktrace or error message from your provider did you see?
See above

## Additional Data
* ***Serverless Framework Version you're using***: 1.14.0
* ***Operating System***: MacOS
* ***Stack Trace***: see above
* ***Provider Error messages***:

**Comment by yyamano** [comment]
I have similar issue. Extracting jar failed. It's easy to reproduce it.

```
% sls create -t aws-java-maven -p aws-java-maven
[snip]
% cd aws-java-maven
% mvn package
[snip]
% env SLS_DEBUG='*' sls invoke local -f hello --docker
Serverless: Load command interactiveCli
Serverless: Load command config
Serverless: Load command config:credentials
Serverless: Load command config:tabcompletion
Serverless: Load command config:tabcompletion:install
Serverless: Load command config:tabcompletion:uninstall
Serverless: Load command create
Serverless: Load command install
Serverless: Load command package
Serverless: Load command deploy
Serverless: Load command deploy:function
Serverless: Load command deploy:list
Serverless: Load command deploy:list:functions
Serverless: Load command invoke
Serverless: Load command invoke:local
Serverless: Load command info
Serverless: Load command logs
Serverless: Load command metrics
Serverless: Load command print
Serverless: Load command remove
Serverless: Load command rollback
Serverless: Load command rollback:function
Serverless: Load command slstats
Serverless: Load command plugin
Serverless: Load command plugin
Serverless: Load command plugin:install
Serverless: Load command plugin
Serverless: Load command plugin:uninstall
Serverless: Load command plugin
Serverless: Load command plugin:list
Serverless: Load command plugin
Serverless: Load command plugin:search
Serverless: Load command config
Serverless: Load command config:credentials
Serverless: Load command rollback
Serverless: Load command rollback:function
Serverless: Load command upgrade
Serverless: Load command uninstall
Serverless: Load command login
Serverless: Load command logout
Serverless: Load command generate-event
Serverless: Load command test
Serverless: Load command dashboard
Serverless: Load command output
Serverless: Load command output:get
Serverless: Load command output:list
Serverless: Load command param
Serverless: Load command param:get
Serverless: Load command param:list
Serverless: Load command studio
Serverless: Load command dev
Serverless: Invoke invoke:local
Serverless: Invoke package
Serverless: Invoke aws:common:validate
Serverless: Invoke aws:common:cleanupTempDir
Serverless: Packaging service...
Serverless: Invoke aws:package:finalize
Serverless: Invoke aws:common:moveArtifactsToPackage
 
  Error --------------------------------------------------
 
  Error: ENOENT: no such file or directory, open '/private/tmp/aws-java-maven/.serverless/invokeLocal/artifact/META-INF/'
      at /opt/pkg/lib/node_modules/serverless/lib/plugins/aws/invokeLocal/index.js:452:19
  From previous event:
      at AwsInvokeLocal.invokeLocalDocker (/opt/pkg/lib/node_modules/serverless/lib/plugins/aws/invokeLocal/index.js:451:8)
      at AwsInvokeLocal.invokeLocal (/opt/pkg/lib/node_modules/serverless/lib/plugins/aws/invokeLocal/index.js:195:19)
  From previous event:
      at Object.invoke:local:invoke [as hook] (/opt/pkg/lib/node_modules/serverless/lib/plugins/aws/invokeLocal/index.js:51:57)
      at /opt/pkg/lib/node_modules/serverless/lib/classes/PluginManager.js:483:55
  From previous event:
      at PluginManager.invoke (/opt/pkg/lib/node_modules/serverless/lib/classes/PluginManager.js:483:22)
      at /opt/pkg/lib/node_modules/serverless/lib/classes/PluginManager.js:518:24
  From previous event:
      at PluginManager.run (/opt/pkg/lib/node_modules/serverless/lib/classes/PluginManager.js:518:8)
      at /opt/pkg/lib/node_modules/serverless/lib/Serverless.js:133:33
  From previous event:
      at Serverless.run (/opt/pkg/lib/node_modules/serverless/lib/Serverless.js:120:74)
      at /opt/pkg/lib/node_modules/serverless/scripts/serverless.js:54:26
      at processImmediate (internal/timers.js:456:21)
      at process.topLevelDomainCallback (domain.js:137:15)
  From previous event:
      at Object.<anonymous> (/opt/pkg/lib/node_modules/serverless/scripts/serverless.js:54:4)
      at Module._compile (internal/modules/cjs/loader.js:1200:30)
      at Object.Module._extensions..js (internal/modules/cjs/loader.js:1220:10)
      at Module.load (internal/modules/cjs/loader.js:1049:32)
      at Function.Module._load (internal/modules/cjs/loader.js:937:14)
      at Module.require (internal/modules/cjs/loader.js:1089:19)
      at require (internal/modules/cjs/helpers.js:73:18)
      at Object.<anonymous> (/opt/pkg/lib/node_modules/serverless/bin/serverless.js:41:1)
      at Module._compile (internal/modules/cjs/loader.js:1200:30)
      at Object.Module._extensions..js (internal/modules/cjs/loader.js:1220:10)
      at Module.load (internal/modules/cjs/loader.js:1049:32)
      at Function.Module._load (internal/modules/cjs/loader.js:937:14)
      at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
      at internal/main/run_main_module.js:17:47
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com
 
  Your Environment Information ---------------------------
     Operating System:          darwin
     Node Version:              14.3.0
     Framework Version:         1.77.1
     Plugin Version:            3.6.18
     SDK Version:               2.3.1
     Components Version:        2.33.0
```

The following simple script doesn't work, so I guess decompress might have an issue.

```
% cat index.js 
const decompress = require("decompress");

decompress("hello-dev.jar", "artifact")
  .then((files) => {
    console.log("extracted!");
  })
  .catch((e) => {
    console.log(e);
  });
% node index.js 
[Error: EISDIR: illegal operation on a directory, open 'artifact/META-INF/'] {
  errno: -21,
  code: 'EISDIR',
  syscall: 'open',
  path: 'artifact/META-INF/'
}
```

**Comment by yhrabovskyinomi** [comment]
same problem

**Comment by dforsl** [comment]
Same problem here (altho on Ubuntu). Without reflecting on whatever other implications there might be, the trick from here does atleast make that issue go away! https://github.com/kevva/decompress/issues/46

Basically, passing a filter to `decompress` assuming only directories end with a forward-slash, due to the lib (or whatever) apparently interpreting the directory as a file:
`decompress(source, destination, {
  filter: file => !file.path.endsWith('/')
})`

**Comment by yyamano** [comment]
I just realized Java zip/jar implementation doesn't preserve file permissions in a zip file.  https://bugs.openjdk.java.net/browse/JDK-6194856
So decompress-unzip can't detect file type correctly.  https://github.com/kevva/decompress-unzip/blob/master/index.js

**Comment by yyamano** [comment]
I'm working on it on https://github.com/yyamano/serverless/tree/issue-6028

**Comment by yyamano** [comment]
@pgrzesik Could you close the issue? 
Already fixed by #9259. Sorry to bother you.

**Comment by pgrzesik** [comment]
Totally, sorry for not closing and thanks for following up @yyamano :bow:
